stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,296 @@
+    reqs = ("tensorflow",) + tuple(map(str, pkg_info.requires()))
+    path = outdir / "requirements.txt"
+    with open(path, "w") as f:
+        f.write("\n".join(reqs))
+    return f"pip:{path}"
+
+
+def _create_stardist_doc(outdir):
+    doc_path = outdir / "README.md"
+    text = (
+        "# StarDist Model\n"
+        "This is a model for object detection with star-convex shapes.\n"
+        "Please see the [StarDist repository](https://github.com/stardist/stardist) for details."
+    )
+    with open(doc_path, "w") as f:
+        f.write(text)
+    return doc_path
+
+
+def _get_stardist_metadata(outdir):
+    metadata, _ = _import()
+    package_data = metadata("stardist")
+    doi_2d = "https://doi.org/10.1007/978-3-030-00934-2_30"
+    doi_3d = "https://doi.org/10.1109/WACV45572.2020.9093435"
+    data = dict(
+        description=package_data["Summary"],
+        authors=list(dict(name=name.strip()) for name in package_data["Author"].split(",")),
+        git_repo=package_data["Home-Page"],
+        license=package_data["License"],
+        dependencies=_create_stardist_dependencies(outdir),
+        cite={"Cell Detection with Star-Convex Polygons": doi_2d,
+              "Star-convex Polyhedra for 3D Object Detection and Segmentation in Microscopy": doi_3d},
+        tags=["stardist", "segmentation", "instance segmentation", "object detection", "tensorflow"],
+        covers=["https://raw.githubusercontent.com/stardist/stardist/master/images/stardist_logo.jpg"],
+        documentation=_create_stardist_doc(outdir),
+    )
+    return data
+
+
+# TODO factor that out (it's the same as in csbdeep.base_model)
+def _get_weights_name(model, prefer="best"):
+    # get all weight files and sort by modification time descending (newest first)
+    weights_ext = ("*.h5", "*.hdf5")
+    weights_files = chain(*(model.logdir.glob(ext) for ext in weights_ext))
+    weights_files = reversed(sorted(weights_files, key=lambda f: f.stat().st_mtime))
+    weights_files = list(weights_files)
+    if len(weights_files) == 0:
+        raise ValueError("Couldn't find any network weights (%s) to load." % ', '.join(weights_ext))
+    weights_preferred = list(filter(lambda f: prefer in f.name, weights_files))
+    weights_chosen = weights_preferred[0] if len(weights_preferred) > 0 else weights_files[0]
+    return weights_chosen.name
+
+
+# TODO we may need to permute axes for images with channel as well
+def _expand_dims(x, axes):
+    n_expand = len(axes) - x.ndim
+    assert n_expand in (0, 1, 2)
+    if n_expand == 0:
+        return x
+
+    # batch should always be first
+    assert axes[0] == "b"
+    if n_expand == 1:
+        return x[None]
+
+    # channel first or channel last
+    assert axes[1] == "c" or axes[-1] == "c"
+    if axes[1] == "c":
+        expander = np.s_[None, None]
+    else:
+        expander = np.s_[None, ..., None]
+    expanded = x[expander]
+    assert expanded.ndim == len(axes)
+    return expanded
```

Comment:
This function is expanding the dims of `x` to the provided `axes`, right?
As you noted in your TODO comment, it will not work in general, it's just guessing correctly in most cases.

We already have [a function to do this](https://github.com/CSBDeep/CSBDeep/blob/b0d2f5f344ebe65a9b4c3007f4567fe74268c813/csbdeep/utils/utils.py#L213-L250) in CSBDeep, but it needs to know the axes of `x` to do its job. E.g.
`move_image_axes(x, 'YX', 'SYXC', adjust_singletons=True)` (`b` in your code is the same as `S` in ours).

Reply:
Essentially, we need to know the axes of the test input image(s).

Reply:
I agree that it would be better to use `move_image_axes` and I also think that it's ok to add the input axes as another function argument. I added this in 0178582, but apparently did something wrong, because the tests now fail with:
```
ValueError: image with shape (512, 512) and axes YX not compatible with target axes SYXC.
```
Could you have a look at that and see how to fix it, @uschmidt83?

Reply:
Argh, I'm also refactoring a lot of code at the moment. I'll merge everything.

Reply:
> Argh, I'm also refactoring a lot of code at the moment. I'll merge everything.

Oh, sorry I wasn't aware. I will stop working on it for now; let me know when all changes are done on your end ;).

Reply:
(and feel free to revert my latest commit if that makes merging easier for you)

Reply:
I might manually copy some bits of your commit and then revert it.

Reply:
I just pushed my refactored code.